DEVELOPER_GUIDE.md

```diff
@@ -0,0 +1,178 @@
+
+All of the above more or less maps to this code:
+
+```js
+if (parser.eat('!--')) {
+	const data = parser.read_until(regex_closing_comment);
+	parser.eat('-->', true);
+
+	parser.append({
+		type: 'Comment',
+		start,
+		end: parser.index,
+		data
+	});
+
+	return;
+}
+```
+
+If the parser doesn't enter this `if`, it will check for all the other language constructs using different strategies to read the information that is needed in the AST (an HTML element for example will need the name, the list of arguments, the fragment etc).
+
+If you want to familiarize yourself with the `svelte` AST, you can go [to the playground](https://svelte.dev/playground), write your `svelte` component and open the `AST Output` tab. This will not only show you the AST of the component but also provide you with hover functionality that will highlight each section of the component when you hover over a section of the AST (and vice versa).
+
+![svelte playground showing the AST tab and the hover functionality](assets/developer-guide/ast.png)
+
+## Phase 2: Analysis
+
+Once we have a AST we need to perform analysis on it...during this phase we will collect information about which variables are used, where are they used, if they are stores etc etc. This information will be later used during the third phase to properly transform and optimizing your component (for example if you declare a stateful variable but never reassign to it or never use it in a reactive context we will not bother with creating a stateful variable at all).
+
+The very first thing to do is to create the scopes for every variable. What this operation does is to create a map from a node to a specific set of references, declarations and declarators. This is useful because if you have a situation like this
+
+```svelte
+<script module>
+	let count = $state(0);
+</script>
+
+<script>
+	let count = $state(0);
+
+	function log(count) {
+		console.log(count);
+	}
+
+	function increase() {
+		count++;
+	}
+</script>
+
+{count}
+```
+
+depending on where you read `count` it will refer to a different variable that has been shadowed. The `count` in the template and in `increase` refers to the `count` declared in instance script, while the one in the `log` function will refer to it's argument.
```

Comment:
```suggestion
Depending on where you read `count` it will refer to a different variable that has been shadowed. The `count` in the template and in `increase` refers to the `count` declared in instance script, while the one in the `log` function will refer to it's argument.
```